### DISCUSSION TITLE
Custom hook for mutation not working

### DISCUSSION THREAD

**Discussion by lagroms** [TanStack/query#5862] [Q&A] [answered]
Hello ! 

I'm trying to implement a custom hook for mutation.
The idea for the custom hook is to take any endpoint,  execute it,  and then execute the successFn when onSuccess: 

```javascript 

import { useMutation } from "@tanstack/react-query";
import useApi from "./useApi";
import { toast } from "react-toastify";

const useMutate = (endpoint, successFn, errorMessage = undefined) => {
    const api = useApi(endpoint, true, true);

    return useMutation({
        mutationFn: (...args) => {
            return api.request(...args);
        },
        onSuccess: (data) => {
            successFn(data);
        },
        onError: () => toast.error(errorMessage ?? "Something Wrong Happened!"),
    });
};

export default useMutate;
```

and then calling the customhook like this in my component 

```javascript
const { mutate: mutatePin } = useMutate(boardsApi.togglePostPin, onRefresh);

return (
    <PostItem onTogglePin={() => mutatePin(post.board.uuid, post.uuid)} />)
)

```

However I don't understand why it's throwing me an error :
![Screenshot 2023-08-10 at 4 56 39 PM](https://github.com/TanStack/query/assets/88441919/fef178d5-9cb1-431e-a33a-9ebf9f94e99b)

### EDIT: well I'm dumb... I just forgot to add the `.js` extension on my custom hook

**Comment by fredrik-andersson-yoin** [comment, marked as answer]
I think that if there are multiple arguments to the mutation function they must be supplied as an object.

`mutationFn: (variables: TVariables) => Promise<TData>`
and
`mutate: (variables: TVariables, { onSuccess, onSettled, onError }) => void`
from: https://tanstack.com/query/v4/docs/react/reference/useMutation

**Reply by lagroms** [reply]
Well I didn't know I have this issue, but thanks to you it's fixed as well 🤣

**Reply by TkDodo** [reply]
https://tkdodo.eu/blog/mastering-mutations-in-react-query#mutations-only-take-one-argument-for-variables
